pkg/keycloak/keycloak.go

```
@@ -2,59 +2,228 @@ package keycloak
+		}
+
+		// NOTE: This may need to be improved in order to abstract it for the shared service lib
+		if svcClassExtMetadata.ServiceName == KEYCLOAK_SERVICE_NAME {
+			return &sc, nil
 		}
 	}
 
-	return nil
+	return nil, errors.Wrap(err, "Failed to find service class")
+}
+
+func (h *Handler) createAdminCredentials(namespace string) (string, error) {
+	adminCredentialsSecret := &corev1.Secret{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "Secret",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:    namespace,
+			GenerateName: "keycloak-admin-cred-",
+		},
+		StringData: map[string]string{
+			"ADMIN_USERNAME": "admin",
+			"ADMIN_PASSWORD": h.generatePassword(),
+		},
+		Type: "Opaque",
+	}
+
+	if err := sdk.Create(adminCredentialsSecret); err != nil {
+		return "", errors.Wrap(err, "Failed to create secret for the admin credentials")
+	}
+
+	return adminCredentialsSecret.GetName(), nil
+}
+
+func (h *Handler) createServiceInstance(namespace string, parameters []byte, sc v1beta1.ClusterServiceClass) v1beta1.ServiceInstance {
+	return v1beta1.ServiceInstance{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "servicecatalog.k8s.io/v1beta1",
+			Kind:       "ServiceInstance",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:    namespace,
+			GenerateName: sc.Spec.ExternalName + "-",
+		},
+		Spec: v1beta1.ServiceInstanceSpec{
+			PlanReference: v1beta1.PlanReference{
+				ClusterServiceClassExternalName: sc.Spec.ExternalName,
+			},
+			ClusterServiceClassRef: &v1beta1.ClusterObjectReference{
+				Name: sc.Name,
+			},
+			ClusterServicePlanRef: &v1beta1.ClusterObjectReference{
+				Name: "default",
+			},
+			Parameters: &runtime.RawExtension{Raw: parameters},
+		},
+	}
+}
+
+func (h *Handler) generatePassword() string {
+	generatedPassword, err := uuid.NewRandom()
+	if err != nil {
+		fmt.Println("Error generating password, setting to default value", err)
```

I think just return the error, rather than the default password, here.